tests/testthat/test-prioritize_dt.R

```diff
@@ -0,0 +1,69 @@
+library(data.table)
+library(testthat)
+
+
+# Test basic scenario -----------------------------------------------------
+
+# set up test input data.table
+input_dt <- CJ(
+  location = "USA", year = c(2000, 2010), age_start = seq(0, 80, 5),
+  report = c(2015, 2020), # numeric variable to prioritize with
+  method = c("A", "B"), # categorical variable to prioritize with
+  value = 1
+)
+id_cols <- setdiff(names(input_dt), "value")
+setkeyv(input_dt, id_cols)
+
+rank_order <- list(
+  method = c("B", "A"),
+  report = -1
+)
+
+expected_dt <- copy(input_dt)
+expected_dt[method == "B" & report == 2020, priority := 1L]
+expected_dt[method == "B" & report == 2015, priority := 2L]
+expected_dt[method == "A" & report == 2020, priority := 3L]
+expected_dt[method == "A" & report == 2015, priority := 4L]
+setkeyv(expected_dt, c(id_cols, "priority"))
+
+testthat::test_that("prioritization of data works", {
+  output_dt <- prioritize_dt(
+    dt = input_dt,
+    rank_by_cols = c("location", "year"),
+    rank_order = rank_order
+  )
+  testthat::expect_identical(output_dt, expected_dt)
+})
+
+# Rank order missing some categorical factor levels -----------------------
+
+# same input as above but with method 'C' & 'D' added to input
+
+# set up test input data.table
+input_dt <- CJ(
+  location = "USA", year = c(2000, 2010), age_start = seq(0, 80, 5),
+  report = c(2015, 2020), # numeric variable to prioritize with
+  method = c("A", "B", "C", "D"), # categorical variable to prioritize with
+  value = 1
+)
+id_cols <- setdiff(names(input_dt), "value")
+setkeyv(input_dt, id_cols)
+
+expected_dt <- copy(input_dt)
+expected_dt[method == "B" & report == 2020, priority := 1L]
+expected_dt[method == "B" & report == 2015, priority := 2L]
+expected_dt[method == "A" & report == 2020, priority := 3L]
+expected_dt[method == "A" & report == 2015, priority := 4L]
+expected_dt[method %in% c("C", "D"), priority := NA]
+setkeyv(expected_dt, c(id_cols, "priority"))
+
+
+testthat::test_that("prioritization of data works", {
+  output_dt <- prioritize_dt(
+    dt = input_dt,
+    rank_by_cols = c("location", "year"),
+    rank_order = rank_order,
+    quiet = TRUE
+  )
+  testthat::expect_identical(output_dt, expected_dt)
+})
```

Comment:
Similar to Katie's suggestions above, would it be helpful to add a test where there are two sources with the same method for the same year to make sure it errors out how we would expect?

Reply:
Agreed, I think what you're describing is what I added on line 40. Including the details below and let me know if you think this doesn't test that scenario. 

where the test input has two separate reports with two methods each but the `rank_order` only includes prioritization of method which leads to duplicate prioritization of 
```
> input_dt[year == 2000 & age_start == 0]
   location year age_start report method value
1:      USA 2000         0   2015      A     1
2:      USA 2000         0   2015      B     1
3:      USA 2000         0   2020      A     1
4:      USA 2000         0   2020      B     1
> rank_order["method"]
$method
[1] "B" "A"
```

The error looks like 
```
> prioritize_dt(
+       dt = input_dt,
+       rank_by_cols = c("location", "year"),
+       unique_id_cols = c("location", "year", "age_start"),
+       rank_order = rank_order["method"]
+     )
Error in prioritize_dt(dt = input_dt, rank_by_cols = c("location", "year"),  : 
  Specified `rank_by_cols`, `rank_order` & returned `priority` do not uniquely identify each row of `dt`.
	- use `warn_non_unique_priority=TRUE` to return `dt` and run demUtils::identify_non_unique_dt
	 with `id_cols = c('location', 'year', 'age_start', 'priority')`
    location year age_start priority
 1:      USA 2000         0        1
 2:      USA 2000         0        2
 3:      USA 2000         5        1
 4:      USA 2000         5        2
```

If we do the same call with `warn_non_unique_priority = TRUE`, the output looks below where the same priority is assigned to the different reports of the same method. 
```
     location year age_start report method value priority
  1:      USA 2000         0   2015      A     1        2
  2:      USA 2000         0   2015      B     1        1
  3:      USA 2000         0   2020      A     1        2
  4:      USA 2000         0   2020      B     1        1
```